src/bun.js/node/util/validators.zig

```diff
@@ -0,0 +1,233 @@
+
+        if (!options.allow_array and value.jsType().isArray()) {
+            try throwErrInvalidArgType(globalThis, name_fmt, name_args, "object", value);
+        }
+
+        if (!value.isObject() and (!options.allow_function or !value.jsType().isFunction())) {
+            try throwErrInvalidArgType(globalThis, name_fmt, name_args, "object", value);
+        }
+    }
+}
+
+pub fn validateArray(globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype, comptime min_length: ?i32) !void {
+    if (!value.jsType().isArray()) {
+        const actual_type = getTypeName(globalThis, value);
+        try throwErrInvalidArgTypeWithMessage(globalThis, "\"" ++ name_fmt ++ "\" property must be an instance of Array, got {s}", name_args ++ .{actual_type});
+    }
+    if (comptime min_length != null) {
+        if (value.length < min_length) {
+            try throwErrInvalidArgValue(globalThis, name_fmt ++ " must be longer than {d}", name_args ++ .{min_length});
+        }
+    }
+}
+
+pub fn validateStringArray(globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype) !void {
+    try validateArray(globalThis, value, name_fmt, name_args, null);
+    var i: usize = 0;
+    var iter = value.arrayIterator(globalThis);
+    while (iter.next()) |item| {
+        if (!item.isString()) {
+            try throwErrInvalidArgType(globalThis, name_fmt ++ "[{d}]", name_args ++ .{i}, "string", value);
+        }
+        i += 1;
+    }
+}
+
+pub fn validateBooleanArray(globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype) !void {
+    try validateArray(globalThis, value, name_fmt, name_args, null);
+    var i: usize = 0;
+    var iter = value.arrayIterator(globalThis);
+    while (iter.next()) |item| {
+        if (!item.isBoolean()) {
+            try throwErrInvalidArgType(globalThis, name_fmt ++ "[{d}]", name_args ++ .{i}, "boolean", value);
+        }
+        i += 1;
+    }
+}
+
+pub fn validateFunction(globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype) !void {
+    if (!value.isFunction())
+        try throwErrInvalidArgType(globalThis, name_fmt, name_args, "function", value);
+}
+
+pub fn validateUndefined(globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype) !void {
+    if (!value.isUndefined())
+        try throwErrInvalidArgType(globalThis, name_fmt, name_args, "undefined", value);
+}
+
+pub fn validateStringEnum(comptime T: type, globalThis: *JSGlobalObject, value: JSValue, comptime name_fmt: string, name_args: anytype) !T {
+    const str = value.toBunString(globalThis);
+    inline for (@typeInfo(T).Enum.fields) |enum_field| {
+        if (str.eqlComptime(enum_field.name))
+            return @field(T, enum_field.name);
+    }
+
+    const values_info = comptime blk: {
+        var out: []const u8 = "";
+        for (@typeInfo(T).Enum.fields, 0..) |enum_field, i| {
```

Comment:
What the heck is going on here? Any docs I can read up to understand this?

Reply:
It's a compile-time snippet that concats the enum fields names into a string, to show in the error message,  using compile-time reflection. Made it based on the official zig docs such as [this](https://ziglang.org/documentation/master/#enum:~:text=//%20%40typeInfo%20tells%20us%20the%20field%20count%20and%20the%20fields%20names%3A)